### ISSUE TITLE
Mixer module sometimes hangs calling pygame.quit() on Virtual Machine/ArchLinux (329)

### ISSUE THREAD

**Issue by GalacticEmperor1** [pygame-community/pygame-ce#344]
#  **Issue &numero;329 opened by *[illume](https://github.com/illume)* at 2017-03-05 21:43:33**

Originally reported by: **ocket8888 (Bitbucket: [ocket8888](https://bitbucket.org/ocket8888), GitHub: [ocket8888](https://github.com/ocket8888))**

----------------------------------------

Just like the title says, I've never been able to write a pygame script that will exit properly. Even something as simple as

```
# !python

# !/usr/bin/env python3

import pygame
from sys import exit as sysExit

pygame.init()
screen = pygame.display.set_mode((512,512))
screen.fill((255,255,255))

pygame.display.update()

while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.display.update()
            pygame.display.quit()
            pygame.quit()
            exit(0)
            sysExit(0)
            # None of these work, code under them is unreachable but the process hangs
```
hangs. More information is available [here](http://stackoverflow.com/questions/42572322/python-3-pygame-programs-will-not-quit)

----------------------------------------
- Bitbucket: https://bitbucket.org/pygame/pygame/issue/329

Steps to proceed with this:

- [ ] Try the code in the above with pygame 2.0.0.dev10 running in a linux virtual machine.
- [ ] If it is still broken try the code above replacing `pygame.init()` & `pygame.display.init()`

**Related Docs:** https://www.pygame.org/docs/ref/pygame.html
<hr>

#  Comments

# # #  **[illume](https://github.com/illume)* commented at 2017-03-07 21:17:45*

*Original comment by* **ocket8888 (Bitbucket: [ocket8888](https://bitbucket.org/ocket8888), GitHub: [ocket8888](https://github.com/ocket8888))**:

----------------------------------------

After hanging out in the irc, user DR0ID_ helped me figure out that this issue is caused specifically by the pygame.mixer module, and he thinks it's probably caused by the fact that I'm on a virtual machine which could be causing driver issues.

<hr>

# # #  **[illume](https://github.com/illume)* commented at 2017-03-13 21:37:47*

*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

It appears to exit normally on Linux (Fedora 25) running on real hardware, so issues relating to the VM sounds plausible.

<hr>

# # #  **[illume](https://github.com/illume)* commented at 2017-03-13 21:58:59*

*Original comment by* **ocket8888 (Bitbucket: [ocket8888](https://bitbucket.org/ocket8888), GitHub: [ocket8888](https://github.com/ocket8888))**:

----------------------------------------

I have confirmed that the 64-bit Linux VM on my desktop will exit the same script just fine, so the architecture could be related to the issue; the VM with the problem is 32-bit. There aren't any significant differences in the OS's themselves, same version of VirtualBox, same versions of pygame and python3.

<hr>

# # #  **[adicarlo](https://github.com/adicarlo)* commented at 2017-09-22 00:09:06*

From comments, it looks like this can only be reproduced under a 32-bit virtualbox instance.  Which kernel is in use?

It might be helpful to run the process under strace.  Specifically, we could look at what exactly is hanging up during cleanup.

<hr>

# # #  **[surculus12](https://github.com/surculus12)* commented at 2018-05-29 16:50:47*

I am experiencing this issue in a 64-bit virtualbox instance (`Version 5.2.12 r122591 (Qt5.6.2)`) and on a Windows 10 Enterprise host (`10.0.15063 Build 15063`).
```
$ uname -a
Linux surculus12-VirtualBox 4.13.0-39-generic # 44~16.04.1-Ubuntu SMP Thu Apr 5 16:43:10 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.4 LTS
Release:        16.04
Codename:       xenial
```

Example of program run (`pygame 1.9.3`, `python 3.6.5`):
```
def foo():
    pygame.init()

    #  Create window to facilitate QUIT event
    screen = pygame.display.set_mode((640, 480,))

    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()  #  Hangs here
                sys.exit()
```

The program hangs inside of `pygame.quit()`. 

Bypassing most pygame initialization via `pygame.display.init()` causes graceful exit.
<hr>

# # #  **[jackjt8](https://github.com/jackjt8)* commented at 2018-06-07 12:28:50*

I'm experiencing this issue, or at least a similar one, trying to exit Pygame on a Raspberry Pi Zero running Raspbrain Stretch (default python installs). 

`pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.14.34+ # 1110 Mon Apr 16 14:51:42 BST 2018 armv6l GNU/Linux
pi@raspberrypi:~ $ python -V
`

`Python 2.7.13
Python 3.5.3 
Pygame 1.9.3`

With Python2 `pygame.quit()` causes it to hang. For whatever reason, SSH'ing and using `sudo kill id`, `sudo kill -9 id`, and `sudo killall python` are ineffective. Likewise, no shutdown method will work if the script is still running. I can only pull the power which may corrupt my SD card. Running script from SSH and then using crtl+c via SSH will "lock up" that SSH window and I need to create a new SSH session to do anything. Switching `pygame.init()` to `pygame.display.init()` does solve the issue for me.

Running the program under Python3 produces a `Segmentation fault` and a black framebuffer. Rerunning the script it will then draw a white box with a movable mouse pointer but it will not auto exit, using crtl+c it will only remove the top half of the box and won't clear the framebuffer. Running script from SSH and using crtl+c via SSH will break the framebuffer as above, but the SSH window will still be usable. Switching `pygame.init()` to `pygame.display.init()` does not solve the issue with Python3 and it acts the same as before.

tl;dr Python2 pygam.exit() is fixed by switching `pygame.init()` to `pygame.display.init()`. Python3 pygame throws a `Segmentation fault` on first run, produces a white box with mouse pointer on all other runs and doesn't auto exit.

<hr>

# # #  **[Bottersnike](https://github.com/Bottersnike)* commented at 2018-06-24 19:48:08*

I have the same issue on Arch Linux x64. My current solution is either to just init the things I need and then `.quit` them and only them or use `os._exit(0)`. I'm using PulseAudio, but not a VM.
<hr>

# # #  **[nudelbrot](https://github.com/nudelbrot)* commented at 2019-01-11 23:05:05*

Can confirm the issue, also on Arch Linux x64; not a VM.
@jackjt8's solution by switching from `pygame.init()` to `pygame.display.init()` solves the problem, also when using Python3.
<hr>

# # #  **[paunis-zz](https://github.com/paunis-zz)* commented at 2019-07-19 11:33:22*

I have the same issue but it's intermitent. Sometimes my app closes properly, sometimes it hangs. I'm running it in Python 3 on a Linux virtual machine.
<hr>

# # #  **[illume](https://github.com/illume)* commented at 2019-07-19 11:42:44*

@paunis I wonder which version of pygame you are using? Have you tried `pygame==2.0.0.dev3`?
<hr>

# # #  **[paunis-zz](https://github.com/paunis-zz)* commented at 2019-07-19 12:28:49*

> > 
> @paunis I wonder which version of pygame you are using? Have you tried `pygame==2.0.0.dev3`?

Latest 1.9.6:
Name: pygame
Version: 1.9.6
Summary: Python Game Development
Home-page: https://www.pygame.org
Author: A community project.
Author-email: pygame@pygame.org
License: LGPL
 
<hr>

# # #  **[MyreMylar](https://github.com/MyreMylar)* commented at 2020-05-12 09:17:56*

So to summarise this issue. This code:

    import sys
    import pygame
    
    pygame.init()
    
    screen = pygame.display.set_mode((640, 480,))
    
    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

Causes pygame to hang on some system setups. 

I've just tried it on Windows in 1.9.6 and 2.0.0.dev8 and it closes fine in both. There were also one report of the issue being intermittent so I also made this program which will init() and quit() pygame 25 times over 50 seconds:

    import sys
    import pygame
    
    for _ in range(25):
        pygame.init()
    
        screen = pygame.display.set_mode((640, 480))
    
        clock = pygame.time.Clock()
        clock.tick()
        timer = 0.0
    
        running = True
        while running:
            timer += clock.tick()
            if timer > 500:
                pygame.event.post(pygame.event.Event(pygame.QUIT))
                timer = 0.0
    
            pygame.display.update()
    
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    running = False
    
    sys.exit()

Which also works fine for me on Windows in pygame 1.9.6, and pygame 2.0.0.dev8.

Since the majority of the reports were from Virtual Machines of some kind I think that is probably the main area of fruitful research. It also appears to be strongly related to the mixer module, given several reports of people fixing it by not calling `pygame.init()` but just calling `pygame.display.init()`.

As such I'll label it with that information and rename it.
<hr>

# # #  **[MyreMylar](https://github.com/MyreMylar)* commented at 2020-05-16 13:16:17*

Move To Do to first post.
<hr>

# # #  **[bmillwood](https://github.com/bmillwood)* commented at 2020-08-24 16:47:13*

I've seen this already exhibited by just:

```
# !/usr/bin/env python3
import pygame

pygame.mixer.init()
print('quitting now')
pygame.mixer.quit()
print('finished quitting')
```

(immediately printing "quitting now" but never "finished quitting"). FWIW ctrl-C and SIGTERM are ineffective but ctrl-\ (SIGQUIT) does the trick.

For me, it hangs when I'm running `fluidsynth` (a software synthesizer I was running to provide a midi device to do midi stuff with), and interestingly, the hang ends (the program exits normally, printing "finished quitting") when I stop `fluidsynth` (I'm just running `fluidsynth --server --audio-driver=alsa -o audio.alsa.device=hw:0 soundfont.sf2` in the foreground and ctrl-C-ing it when I'm done).

Can strace if you give me some particular syscalls to trace -- the unfiltered output is too much for me to review.

I'm on 1.9.6.
<hr>

# # #  **[MyreMylar](https://github.com/MyreMylar)* commented at 2020-08-25 06:43:06*

Hmm. Do you mean that you are running fluidsynth separately independent from whatever you are doing with pygame - or that you are using fluidsynth specifically in order to improve pygame audio?

I believe that SDL (which pygame is built on) also uses/can use fluidsynth to play audio instead of Timidity. This post from way back in 2011 suggest it will be used by SDL first if it is available before falling back on timidity:

https://discourse.libsdl.org/t/patch-fluidsynth-backend-for-sdl-mixer/18070/15

Though that is from 2011/2012 so the situation may have evolved since. Perhaps some code somewhere along the line is waiting for fluidsynth or Timidity to report that it has shut down and it isn't getting shut down (perhaps because SDL didn't start it, whichever one is being waited for) and so it waits forever.

It's possible that this may be magically resolved by the slow migration of changes from SDL Mixer X back into to SDL Mixer? Though that is just wild speculation. I feel like SDL mixer has not traditionally been frequently updated.

(SDL Mixer X: https://github.com/WohlSoft/SDL-Mixer-X)

Though it could be something else entirely. This might be related:

https://github.com/FluidSynth/fluidsynth/pull/478

Though it post dates the first problems we had reported in here.

You can see the pygame C code called when mixer is quit here:

https://github.com/pygame/pygame/blob/1476009b05919c96b22ef9abeabc5a623c552ce7/src_c/mixer.c# L298-L341

No obvious loops that could be hanging there. I guess the only way to take it further code wise would be building an SDL C application, on a linux system that has the issue, one that essentially does the same thing pygame is doing. Then see if it still hangs on running the quit function, and then from there tracking it down to a specific call to SDL.

Looks like it would be a choice of:

Mix_HaltMusic()
Mix_FreeMusic()
Mix_CloseAudio()
SDL_QuitSubSystem(SDL_INIT_AUDIO)

Searching Github some on those function calls, I found this thread:
https://github.com/wesnoth/wesnoth/issues/3588

Which sounds very similar (SDL hanging while another audio using program is running) and is highly suggestive that the problem is likely to lie with Mix_CloseAudio().
<hr>

# # #  **[bmillwood](https://github.com/bmillwood)* commented at 2020-08-27 15:04:32*

I'm playing pygame midi audio through fluidsynth (without it, I don't hear anything at all). I believe I don't have timidity installed.

Indeed it does seem like this C file reproduces my problem:

```c
# include "SDL.h"
# include "SDL_mixer.h"

int main(int argc, char *argv[])
{
  int ret;
  ret = SDL_Init(SDL_INIT_AUDIO);
  if(ret < 0) {
    return 1;
  }
  
  ret = Mix_OpenAudio(MIX_DEFAULT_FREQUENCY, MIX_DEFAULT_FORMAT, 2, 1024);
  if(ret < 0) {
    return 2;
  }

  printf("before Mix_CloseAudio\n");
  Mix_CloseAudio();
  printf("after Mix_CloseAudio\n");
  SDL_Quit();
  return 0;
}

```

(i.e. it hangs on "before Mix_CloseAudio" until I stop fluidsynth).
<hr>

# # #  **[rodetroner](https://github.com/rodetroner)* commented at 2021-07-01 20:48:58*

I was able to reproduce the bug with:
```
import pygame

pygame.mixer.init()
pygame.display.init()
pygame.display.quit()
pygame.mixer.quit()
```
I don't know if this is of any value, but when running `strace python file.py`, where file.py contains the above code, the output freezes at `pselect6(1, [0], NULL, NULL, NULL, NULL`

With the code below, the program terminates correctly.
```
import pygame

pygame.mixer.init()
pygame.mixer.quit()
```
<hr>

# # #  **[ankith26](https://github.com/ankith26)* commented at 2022-01-15 14:35:54*

Reviving old issue, I think we can safely say this is a confirmed bug in `Mix_CloseAudio`, so the next step should be reporting it on `SDL_mixer` github repo with the C code reproducer?
<hr>

# # #  **[MyreMylar](https://github.com/MyreMylar)* commented at 2022-01-15 16:13:22*

I believe it is only with the fluidsynth backend as well, I guess somebody who actually gets the bug should test it (i.e. someone with a linux system and a fluidsynth install)
<hr>

# # #  **[bmillwood](https://github.com/bmillwood)* commented at 2022-01-15 19:24:56*

I posted about my reproduction case in libsdl-org/SDL_mixer# 367 but it's not clear if the other reproductions involve fluidsynth / are affected by the same alsa device thing that I noticed while filing that issue. Would be interested to see other people still affected by this comment on that?

(FWIW, as I mentioned in the SDL ticket, I now have an effective workaround, so this isn't an issue for me personally)
<hr>

# # #  **[mezdahun](https://github.com/mezdahun)* commented at 2022-03-03 08:56:24*

I use pygame to carry out agent-based simulations for which I run a pygame based simulation instance for a given time, save some generated data, then quit, change some parameters in my framework and restart the process. I can reproduce this issue (or closely related) on Ubuntu 20 in an Oracle VM,  in a non-VM installation and in a Docker container where pygame runs in headless mode. The quit method hangs **when I switch user** (when using docker switching user on the host). This might be relevant to fix the issue so I leave the comment here in case it helps fixing the issue or reproduce it. As described before switching to `pygame.display.init()` from `pygame.init()` seemingly solved the problem and I can keep running my simulations with other users using the same machine.

**Comment by MyreMylar** [comment]
I believe this issue was not in pygame but in SDL_mixer. It was raised here: 
https://github.com/libsdl-org/SDL_mixer/issues/367

And thought o have been resolved with this patch: https://github.com/libsdl-org/SDL_mixer/commit/452a55c5d8cd78621114b8aaa258377403797ea7

Our version of SDL_mixer includes this patch by now. So without any further likely follow ups to come I'm closing this for now.
